packages/protobuf/src/create-registry.ts

```diff
@@ -71,9 +48,55 @@ export function createRegistry(
     findService(typeName: string): ServiceType | undefined {
       return services[typeName];
     },
+    findExtensionFor(typeName: string, no: number): Extension | undefined {
+      return extensionsByExtendee.get(typeName)?.get(no) ?? undefined;
+    },
+    findExtension(typeName: string): Extension | undefined {
+      return extensionsByName.get(typeName) ?? undefined;
+    },
   };
+
+  function addType(type: MessageType | EnumType | ServiceType | Extension) {
+    if ("fields" in type) {
+      if (!registry.findMessage(type.typeName)) {
+        messages[type.typeName] = type;
+        type.fields.list().forEach(addField);
+      }
+    } else if ("methods" in type) {
+      if (!registry.findService(type.typeName)) {
+        services[type.typeName] = type;
+        for (const method of Object.values(type.methods)) {
+          addType(method.I);
+          addType(method.O);
+        }
+      }
+    } else if ("extendee" in type) {
+      if (!extensionsByName.has(type.typeName)) {
+        extensionsByName.set(type.typeName, type);
+        const extendeeName = type.extendee.typeName;
+        if (!extensionsByExtendee.has(extendeeName)) {
+          extensionsByExtendee.set(extendeeName, new Map<number, Extension>());
+        }
+        extensionsByExtendee.get(extendeeName)?.set(type.field.no, type);
+        addType(type.extendee);
+        addField(type.field);
+      }
+    } else {
+      enums[type.typeName] = type;
+    }
+  }
+
+  function addField(field: FieldInfo) {
+    if (field.kind == "message") {
+      addType(field.T);
+    } else if (field.kind == "map" && field.V.kind == "message") {
```

Comment:
I guess the `kind` for group fields will be `"message"`? Or does this need to be expanded to include groups?

Reply:
Yes, a group field is `kind: "message"` with `delimited: true`. Since we generate a message class for a group, it's a decent fit. This will also work nicely with the edition feature `message_encoding = DELIMITED`.